pkg/streamingpromql/operators/and_binary_operation.go

```diff
@@ -0,0 +1,297 @@
+	vectorMatching parser.VectorMatching,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	timeRange types.QueryTimeRange,
+	expressionPosition posrange.PositionRange,
+) *AndBinaryOperation {
+	return &AndBinaryOperation{
+		Left:                     left,
+		Right:                    right,
+		VectorMatching:           vectorMatching,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+		timeRange:                timeRange,
+		expressionPosition:       expressionPosition,
+	}
+}
+
+func (a *AndBinaryOperation) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	leftMetadata, err := a.Left.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer types.PutSeriesMetadataSlice(leftMetadata)
+
+	if len(leftMetadata) == 0 {
+		// We can't produce any series, we are done.
+		return nil, nil
+	}
+
+	rightMetadata, err := a.Right.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer types.PutSeriesMetadataSlice(rightMetadata)
+
+	if len(rightMetadata) == 0 {
+		// We can't produce any series, we are done.
+		return nil, nil
+	}
+
+	groupMap := map[string]*andGroup{}
+	groupKeyFunc := vectorMatchingGroupKeyFunc(a.VectorMatching)
+
+	// Iterate through the left-hand series, and create groups for each based on the matching labels.
+	a.leftSeriesGroups = make([]*andGroup, 0, len(leftMetadata))
+
+	for _, s := range leftMetadata {
+		groupKey := groupKeyFunc(s.Labels)
+		group, exists := groupMap[string(groupKey)] // Important: don't extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+		if !exists {
+			group = &andGroup{lastRightSeriesIndex: -1}
+			groupMap[string(groupKey)] = group
+		}
+
+		group.leftSeriesCount++
+		a.leftSeriesGroups = append(a.leftSeriesGroups, group)
+	}
+
+	// Iterate through the right-hand series, and find groups for each based on the matching labels.
+	a.rightSeriesGroups = make([]*andGroup, 0, len(rightMetadata))
+	outputSeriesCount := 0
+
+	for idx, s := range rightMetadata {
+		groupKey := groupKeyFunc(s.Labels)
+		group, exists := groupMap[string(groupKey)] // Important: don't extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+		if exists {
+			if group.lastRightSeriesIndex == -1 {
+				// First time a right-hand series has matched this group.
+				// We'll return all left-hand series matching this group, so add the count to the running total of output series.
+				outputSeriesCount += group.leftSeriesCount
+			}
+
+			group.lastRightSeriesIndex = idx
+		}
+
+		// Even if there is no matching group, we want to store a nil value here so we know to throw the series away when we read it later.
+		a.rightSeriesGroups = append(a.rightSeriesGroups, group)
+	}
+
+	// Iterate through the left-hand series again, and build the list of output series based on those that matched at least one series on the right.
+	outputSeries := types.GetSeriesMetadataSlice(outputSeriesCount)
```

Comment:
could we not reuse the existing leftMetadata slice?

Reply:
Yep, good point, changed in 72e9979